Carbon forms rgw

nizamial09 opened the pull request:
<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

jmundack commented:
can we add some context around why we are making these changes?

nizamial09 commented:
> can we add some context around why we are making these changes?

@jmundack we are replacing the bootstrap framework with Carbon Design System in upstream. You can check [this presentation](https://tracker.ceph.com/attachments/7207) @ivoalmeida gave a couple of months ago in one of the ceph user dev monthly. We also have a [series of tasks](https://tracker.ceph.com/issues/68249) created for it replace every bootstrap (or other custom frameworks) and move them to carbon.

nizamial09 commented:
Thanks @Pegonzal for testing it.. I'll test the ACL and also the some buttons not being aligned. But regarding the form alignment I have a comment below.

> * The elements in this grous are shifted to the right or placed in the middle of the from while their group titles are left aligned. I think this pattern gives much better clouser making it easier to read and differenciate the groups, by better encapsulating the different fields within a group. While the new version, on the right, is placing everyithing left aligned, making it harder to understand

The reason I went for it is because the form alignment guideline in the carbon was mostly saying about all the fields should be vertically aligned and shouldn't vary from its alignment. I didn't see it differ in any of the group section in the form. I remember referring to one of the other product that uses carbon and it looks like they were also following the same pattern. AWS also follows a similar thing but a different way since they use a different framework

![image](https://github.com/user-attachments/assets/ed59a8ba-bf3a-4630-b1a4-89c0c2b60e5e)

nizamial09 commented:
> I'll test the ACL and also the some buttons not being aligned.

@Pegonzal fixed them now. thanks

nizamial09 commented:
@Achintk1491 can you test the ratelimit configurations for both bucket and users and make sure the new changes are working? Also do you have the tracker for e2e testing the rate limit?

Pegonzal commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-user-capability-modal/rgw-user-capability-modal.component.html:
```diff
@@ -1,70 +1,79 @@
+                      formControlName="type"
+                      [invalid]="formGroup.controls.type.invalid && formGroup.controls.type.dirty"
+                      [invalidText]="typeError"
+                      cdRequiredField="Type">
```
Not required since both fields are required for the modal
```suggestion
```
![image](https://github.com/user-attachments/assets/a86ea834-3cc1-4a47-a553-90811e14c1a3)

nizamial09 replied:
and when I say text message I meant the subtext in the form that says `All fields are required unless specified otherwise..`

dnyanee1997 replied:
Thanks @nizamial09
 Form looking really nice, except the buttons for Bucket policy not sure if its carbon design format but i think instead of Clear button we can add icon too.

![Screenshot from 2025-02-25 14-53-16](https://github.com/user-attachments/assets/dbcdf351-8db5-4d55-9e71-de2383b83ec2)
Recently i used cd-help-text in Storage Class Mgmt form. I can updated the same.

nizamial09 replied:
> Bucket policy not sure if its carbon design format but i think instead of Clear button we can add icon too.

i don't think that's a carbon pattern. Its something that we created on our own for the sake of showing these helpers. And the clear button is left that way to match with the other buttons. You are probably right, it looks a little weird here. @Pegonzal I think you were the one who introduced it. WDYT?

and btw, are you sure you are on the latest changes? because I've made some changes after this where I added a little margin at the top of buttons so it looks a little bit more better.


> Recently i used cd-help-text in Storage Class Mgmt form. I can updated the same.

Don't worry. I have added that in this PR. I saw them being used in the Storage Class form and the bucket tiering form that @nmunet works. Changed them both to this.

Pegonzal replied:
Happened to me as well, I found the buttons over there a bit weird as well, but looked trough the carbon guidelines and did not find anything better for this, so I do not know what to suggest other that that.

Btw, did we change the button placement? Is that inentional? I believe the clear button was on the leftside before, not sure if that would be better or if it matters anyways

Pegonzal commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-user-capability-modal/rgw-user-capability-modal.component.html:
```diff
@@ -1,70 +1,79 @@
+                      formControlName="perm"
+                      [invalid]="formGroup.controls.perm.invalid && formGroup.controls.perm.dirty"
+                      [invalidText]="permError"
+                      cdRequiredField="Permission">
```
Same here, no need for the required
```suggestion>
```

nizamial09 replied:
did this too and to other modals as well. thanks

Pegonzal reviewed (approved):
Thanks @nizamial09 tested the PR and works great, leaving some comments for some possible improvements I saw: 

- I think this `context.component` should be carbonized, which works for users and buckets: https://github.com/ceph/ceph/blob/main/src/pybind/mgr/dashboard/frontend/src/app/core/context/context.component.html

![image](https://github.com/user-attachments/assets/8b19c6cc-5f0c-4483-ba3a-24ada3664735)


I see some warnings due to approach used to disable fields:
 ```
 rgw-bucket-form.component.html:192 
  It looks like you're using the disabled attribute with a reactive form directive. If you set disabled to true
  when you set up this control in your component class, the disabled attribute will actually be set in the DOM for
  you. We recommend using this approach to avoid 'changed after checked' errors.

  Example:
  // Specify the `disabled` property at control creation time:
  form = new FormGroup({
    first: new FormControl({value: 'Nancy', disabled: true}, Validators.required),
    last: new FormControl('Drew', Validators.required)
  });

  // Controls can also be enabled/disabled after creation:
  form.get('first')?.enable();
  form.get('last')?.disable();
  ```
  
  - Probably out of the scope of the PR, but the loading here should be carbonized as well
  https://carbondesignsystem.com/patterns/loading-pattern/#full-screen-loading
  
![image](https://github.com/user-attachments/assets/d4f94cbe-0eee-4b67-b3aa-093973cbfaf3)

nizamial09 commented:
Thanks @Pegonzal. updated the PR.

> * I think this `context.component` should be carbonized, which works for users and buckets: https://github.com/ceph/ceph/blob/main/src/pybind/mgr/dashboard/frontend/src/app/core/context/context.component.html

good point, i have added it to the carbon tasks: https://tracker.ceph.com/issues/70159


> I see some warnings due to approach used to disable fields:

that's actually there in the previous forms too and not because of carbon. Its the way some of the fields are disabled using the `disable()` attribute which is all over the forms. Its probably a bigger cleanup. Opened a tracker: https://tracker.ceph.com/issues/70160




> * Probably out of the scope of the PR, but the loading here should be carbonized as well

opened one for this too: https://tracker.ceph.com/issues/70161

Achintk1491 commented:
> @Achintk1491 can you test the ratelimit configurations for both bucket and users and make sure the new changes are working? Also do you have the tracker for e2e testing the rate limit?

No @nizamial09 I don't have a separate tracker for e2e testing of rate limit.

nizamial09 commented:
> > @Achintk1491 can you test the ratelimit configurations for both bucket and users and make sure the new changes are working? Also do you have the tracker for e2e testing the rate limit?
> 
> No @nizamial09 I don't have a separate tracker for e2e testing of rate limit.

can you create one? once the feature works are complete we should extend the e2e

nizamial09 commented:
jenkins test api

nizamial09 commented:
jenkins retest this please

nizamial09 commented:
jenkins test make check arm64
